drivers/can/can_numaker.c

```diff
@@ -0,0 +1,316 @@
+/*
+ * Copyright (c) 2023 Nuvoton Technology Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nuvoton_numaker_canfd
+
+#include <zephyr/drivers/reset.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/drivers/can.h>
+#include <zephyr/drivers/can/can_mcan.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/clock_control/clock_control_numaker.h>
+#include <zephyr/logging/log.h>
+#include <soc.h>
+#include <NuMicro.h>
+
+LOG_MODULE_REGISTER(can_numaker, CONFIG_CAN_LOG_LEVEL);
+
+/* Implementation notes
+ * 1. Use Bosch M_CAN driver (m_can) as backend
+ *    NOTE: m_can only supports single configuration of max element numbers for all instances
+ * 2. Need to modify can_numaker_get_core_clock() for new SOC support
+ */
+
+struct can_numaker_config {
+	mm_reg_t   canfd_base;
+	mem_addr_t  mrba;
+	mem_addr_t  mram;
+	const struct reset_dt_spec reset;
+	uint32_t    clk_modidx;
+	uint32_t    clk_src;
+	uint32_t    clk_div;
+	const struct device *clk_dev;
+	void (*irq_config_func)(const struct device *dev);
+	const struct pinctrl_dev_config *pincfg;
+};
+
+struct can_numaker_data {
+	/* Avoid zero length struct */
+	uint32_t resv0;
+};
+
+static int can_numaker_get_core_clock(const struct device *dev, uint32_t *rate)
+{
+	const struct can_mcan_config *mcan_config = dev->config;
+	const struct can_numaker_config *config = mcan_config->custom;
+	int rc = 0;
+	uint32_t clksrc_rate_idx;
+	uint32_t clksrc_rate;
+	uint32_t clkdiv_divider;
+
+	/* Initialize clock source rate table
+	 * NOTE: Fix the table for new SOC support
+	 */
+	uint32_t clksrc_rate_tab[] = {
+		__HXT,      /* HXT */
+		0,          /* PLLFOUT/2 */
+		0,          /* HCLK */
+		__HIRC      /* HIRC */
+	};
+	clksrc_rate_tab[1] = CLK_GetPLLClockFreq() / 2;
+	clksrc_rate_tab[2] = CLK_GetHCLKFreq();
+
+	/* SOC in support list
+	 * NOTE: * NOTE: Fix the conditional for new SOC support
+	 */
+	if (!IS_ENABLED(CONFIG_SOC_SERIES_M46X)) {
+		LOG_ERR("Not support SOC: " CONFIG_SOC_SERIES);
+		rc = -ENOTSUP;
+		goto cleanup;
+	}
+
+	/* Module clock source rate */
+	clksrc_rate_idx = CLK_GetModuleClockSource(config->clk_modidx);
```

Comment:
why are you not using a switch statement here and just call the correct getter instead of filling a table and than pick one?